bitmark-references/bitmark-node-software/mining.md

```diff
@@ -0,0 +1,123 @@
+
+At bitmark we use `argon2` to compute the hashing, the `argon2` has some benefits such as a memory hard algorithm that provides resistance against GPU and ASIC hardware for computation as well as it provides a hash of 256 bits long.
+
+The `argon2` user-land software provides a wide range of parameters, here there is an example:
+
+```
+printf '%s' 'hello world' | argon2 'hello world' -d -l 32 -m 17 -t 4 -p 1 -r | awk '{for(i=length($1);i>0;i-=2)x=x substr($1,i-1,2);print x}'
+
+f8a17bc25cb53e848e2d09811ade4b8a037f628443661b88611faf5d9a5a1f33
+```
+
+A block hash contains the following data:
+
+1. block record version
+
+    blockchain header version
+
+1. transaction count
+
+    number of transactions in the block
+
+1. merkle tree
+
+    there is an allowable limit of maximum 9999 transactions hashes per block, if the number of verified transactions are above the limit, random transactions are selected.
+
+1. timestamp
+
+    block generated time
+
+1. difficulty
+
+    hashing difficulty
+
+1. nonce
+
+    random string to make the hash fits the difficuty
+
+## Difficulty
+
+It is possible to set the level of difficulty to find a valid hash, higher difficulty means harder to find a valid hash, when a hash meets difficulty, it means the hash value is less than or equal to the difficulty level. Usually the difficulty level is represented by a number.
+
+The hash is considered as a fixed 256 bits value and it is composed by two parts:
+
+    [8 bits] [248 bits]
+    exponent mantissa
+
+Difficulty value is encoded by 64 bits, including 8 bits of exponent and 56 bits of mantissa.
+
+    [56 bits] [8 bits]
+    mantissa  exponent = 64 bits unsigned value
+
+The mantissa bits are normalized, the most significant bit is 1 and can be dropped, leaving the 56 bits to store; these bits are right shifted by exponent +8 from the most significant bit.
+
+The difficulty `One` is defined as `00 ff  ff ff  ff ff  ff ff`, represents 256 bits value `0ff ffff ffff ffff 8000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000`.
+
+The `Two` is the double of difficulty `One`: `01ffffffffffffff`, represents 256 bits value `007fffffffffffffc00000000000000000000000000000000000000000000000`.
+
+## Communication
+
+The `bitmarkd` sends hashing job to `recorderd`. Each `bitmarkd` has a job queue to keep the records of what job has been sent to `recorderd`. If `recorderd` finds a possible hash that might fits the difficulty, it sends the result back to `bitmarkd`, and `bitmarkd` returns the result verified back to `recorderd`.
+
+Here is the message format send from `recorderd` to `bitmarkd`:
+
+1. request
+
+    a string "block.nonce" to represent the type of request
+
+1. job
+
+    `bitmarkd` provides the job's number in the queue
+
+1. packed
+
+    nonce that may makes hash meets difficulty
+
+Here is the message format send from `bitmarkd` to `recorderd`:
+
+1. job
+
+    job number
+
+1. ok
+
+    denote hash is valid or not
+
```

Comment:
Again, examples would be great. Show the communication and break out the parts.

Reply:
updated